zebra/rt_netlink.c

```diff
@@ -2319,29 +2338,14 @@ static int netlink_ipneigh_change(struct nlmsghdr *h, int len, ns_id_t ns_id)
 	ip.ipa_type = (ndm->ndm_family == AF_INET) ? IPADDR_V4 : IPADDR_V6;
 	memcpy(&ip.ip.addr, RTA_DATA(tb[NDA_DST]), RTA_PAYLOAD(tb[NDA_DST]));
 
-	/* Drop some "permanent" entries. */
-	if (ndm->ndm_state & NUD_PERMANENT) {
-		char b[16] = "169.254.0.1";
-		struct in_addr ipv4_ll;
-
-		if (ndm->ndm_family != AF_INET)
-			return 0;
-
-		if (!zif->v6_2_v4_ll_neigh_entry)
-			return 0;
-
-		if (h->nlmsg_type != RTM_DELNEIGH)
-			return 0;
-
-		inet_pton(AF_INET, b, &ipv4_ll);
-		if (ipv4_ll.s_addr != ip.ip._v4_addr.s_addr)
-			return 0;
-
-		if_nbr_ipv6ll_to_ipv4ll_neigh_update(
-			ifp, &zif->v6_2_v4_ll_addr6, true);
+	if (h->nlmsg_type == RTM_DELNEIGH && (ndm->ndm_state & NUD_PERMANENT)) {
+		netlink_handle_5549(ndm, zif, ifp, &ip);
 		return 0;
 	}
 
+	if (h->nlmsg_type == RTM_NEWNEIGH && !(ndm->ndm_state & NUD_VALID))
+		netlink_handle_5549(ndm, zif, ifp, &ip);
+
```

Comment:
5549 naming for the RFC?

Reply:
yes, since the linux kernel makes us jump thru these neighbor hoops for ipv4 prefixes advertised with ipv6 nexthops (rfc5549)

Reply:
I'll also comment the code with why we're taking these actions